### ISSUE TITLE
Separate required Xcode versions for iOS and macOS

### ISSUE THREAD

**Issue by hacker1024** [flutter/flutter#103844]
The required Xcode version was recently bumped to 13. The main reason, for this, though, was for iOS: https://github.com/flutter/flutter/issues/91634.

Should there be different required Xcode versions for iOS and macOS?

**Comment by stuartmorgan** [comment]
Do you have a specific use case here?

This has been discussed before, but there was never a particularly good reason to add the complexity of different requirements.

**Comment by hacker1024** [comment]
I prefer to use macOS Catalina, of which the last supported Xcode version is 12.4. I also have no interest in iOS development, and I am building my app to run on macOS.

The warning in `flutter doctor` is misleading, but everything still works.

**Comment by darshankawar** [comment]
Keeping it open and labeling for further tracking and insights from the team.

**Comment by pkkhandsome** [comment]
Gradually abandoning lower versions of macOS is Apple's strategy.
Although it's disgusting,
I had to update macos and xcode , in order to upload the new app version

**Comment by gspencergoog** [comment]
Thank you for your issue. At this time this is not really the direction in which we would like to take the project. If you are interested in pursuing this further I would encourage you to discuss it in the Discord chat. A design doc with use cases may be a good way to socialise this idea with the rest of the team too.

Our long term intent is to de-duplicate and merge a lot of this iOS/macOS build code, not keep it separate.

**Comment by jmagman** [comment]
When I bump minimum Xcode versions sometimes it _is_ because of necessary Xcode features, like we want our tooling to build for arm64 Macs so need a particular toolchain to get the right linker and clang version.  But more often it's because for iOS we need to require a version of version of clang the same or newer than the one on the engine builders, or else users with bitcode on get App Store submission failures.  And we want to keep bumping the version in the engine builders to get the newest iOS and macOS SDKs.

Bitcode isn't a thing for macOS apps (as an aside the x86->arm64 transition would have been a good use-case for rebuilding those apps, but I guess they relied on Rosetta).  So in those cases it only impacts iOS developers.

> Our long term intent is to de-duplicate and merge a lot of this iOS/macOS build code, not keep it separate.

For my sake, I am 100% for this!  In this particular case though the `Xcode` object is already shared, so I think it should be trivial to keep a minimum required version separate for macOS (we can keep the same "recommended" version, meaning the one we actually test in our CI).  Essentially it would just be another copy of as `isRequiredVersionSatisfactoryForMacOS`
https://github.com/flutter/flutter/blob/8f360c977a618599e6507d6239083a341993d9c6/packages/flutter_tools/lib/src/macos/xcode.dart#L147
and `xcodeMacOSRequiredVersion`
https://github.com/flutter/flutter/blob/8f360c977a618599e6507d6239083a341993d9c6/packages/flutter_tools/lib/src/macos/xcode.dart#L21


The biggest source of confusion will be how to convey the difference in `flutter doctor` since that validator would need to communicate "your version is sufficient for macOS but not iOS".  One easy thing to do would be to keep the doctor check as-is so it looks like an error if you can't build for iOS, but not actually fail during `flutter build macos`.

So actually I think this could be a useful feature so our macOS developers don't have to upgrade their hardware to keep on the newest version of Xcode if they don't necessarily need it.

**Comment by jmagman** [comment]
Reopening, please let me know if you have other concerns though @gspencergoog  🙂

**Comment by jmagman** [comment]
> The warning in `flutter doctor` is misleading, but everything still works as far as I can tell.

And that sounds like a bug...

**Comment by jmagman** [comment]
> > The warning in `flutter doctor` is misleading, but everything still works as far as I can tell.
> 
> And that sounds like a bug...

Indeed, the build macOS code is not checking anything about the Xcode version.  So it essentially works now as I proposed--that the doctor complains, but `flutter build macos` does not.  But of course it should be checking some minimum for macOS as well...

**Comment by gspencergoog** [comment]
No, no concerns.  In Desktop triage we thought that this was asking for a feature that would add a lot of complexity (and duplication). If it doesn't add (much) complexity, then it's fine.

**Comment by stuartmorgan** [comment]
> we thought that this was asking for a feature that would add a lot of complexity

Sorry my earlier comment was misleading; the complexity I meant was the UX problem @jmagman described above.

> But of course it should be checking some minimum for macOS as well...

I'm trying to remember why it's not. I think it may be the case that we never explicitly added anything to the macOS build that actually had a minimum Xcode version requirement (other than needing something with at least the 10.11 SDK, but that's effectively anything.)

**Comment by jmagman** [comment]
> I'm trying to remember why it's not. I think it may be the case that we never explicitly added anything to the macOS build that actually had a minimum Xcode version requirement (other than needing something with at least the 10.11 SDK, but that's effectively anything.)

For macOS we could try to build, and only if it doesn't then check the Xcode version and recommend an upgrade to the iOS recommended version, instead of trying to keep track of what minimum version will build and preemptively stopping without trying.  Ask for forgiveness instead of permission.

For iOS it's trickier since most of our users aren't using bitcode (it's off by default in the template), but if they do turn it on or are using add-to-app then they will hit App Store submission issues on older versions.  And sometimes it just won't build on earlier versions regardless of bitcode.

But because of the bug where we're not enforcing or checking any Xcode version this feature request seems low priority--macOS developers can just ignore the Xcode doctor warning.

**Comment by cbracken** [comment]
We discussed this a bunch in our triage meeting last week when we'd closed it. While I'm very sympathetic to the plight of developers being forced to upgrade Xcode (and transitively, OS) versions, I'm really on the fence about this. 

We'd originally closed this with the view that our infra is already a fairly complex web of dependency versions, and that adding one more variable makes the cartesian product of possible configurations even worse than it already is. Our assumption was that a significant majority of Flutter developers working on a macOS app are also likely to be using Flutter for iOS, and therefore need to be on the same version as iOS regardless. I worry that we're adding additional complexity that impacts testing infra to cover a use-case that benefits a small number of users. I wonder if we have enough opt-in stats from tool usage to identify what fraction of Flutter developers are developing for macOS but not iOS. (/cc @zanderso who may know if we have that data)

I'm fine with keeping this open, but I'm a bit sceptical of the bang for buck here.

**Comment by jmagman** [comment]
> I'm fine with keeping this open, but I'm a bit sceptical of the bang for buck here.

Me too, seems like a low priority but legit feature request.  Users keep Macs a long time since they generally keep working for many years and are really expensive to replace.

**Comment by jmagman** [comment]
> Me too, seems like a low priority but legit feature request.

I've had a few years to think on this 😅  and I now don't think there's a good case for this, particularly now that both will be using Swift Package Manager, and the last Xcode minimum version bumps have been driven by that. As the two platforms converge with SwiftUI, xcodebuild, etc they will likely be required in tandem from now on.
